cpp/src/arrow/io/interfaces.h

```diff
@@ -248,6 +271,12 @@ class ARROW_EXPORT RandomAccessFile : public InputStream, public Seekable {
   /// \return A buffer containing the bytes read, or an error
   virtual Result<std::shared_ptr<Buffer>> ReadAt(int64_t position, int64_t nbytes);
 
+  // EXPERIMENTAL
+  // The caller has to ensure the RandomAccessFile object stays alive
+  // until the future is completed.
+  virtual Result<Future<std::shared_ptr<Buffer>>> ReadAsync(int64_t position,
```

Comment:
Should this API always be `Future<...>` and not `Result<Future<...>>` (even for implementations that don't actually run asynchronously)? Otherwise there are two different failure modes to take into consideration

Reply:
Yes, perhaps it would be simpler indeed.